README.md
PassHash
========

Google Chrome Extension that hashes a passphrase to generate 16-character, base64, (=1024-bit) password on the fly, using no storage or communication. Unlike [LastPass](http://www.lastpass.com), the users' passwords are *not* transmitted, stored centrally, nor stored locally. This is a hashing application that runs client side, in browser, with no login or account. That makes it more of a memory augmentation than a service.

[YouTube Promo](https://www.youtube.com/watch?v=4NxS-SU_6zw)  
[Product Page](http://www.pass-hash.com)  

## Installation

This is source for a Chrome Plugin. It can be downloaded for free from the [Chrome App Store](https://chrome.google.com/webstore/detail/passhash/obghkgeodnccgepekkbomhijojbodfmn). Alternatively, to build it yourself, follow these steps:  

1. Download all repository files in a single directory.  
2. Obtain copies of [jQuery](http://code.jquery.com/jquery-1.11.1.min.js) and QTip2 ([JS](http://cdn.jsdelivr.net/qtip2/2.2.1/jquery.qtip.min.js) + [CSS](http://cdn.jsdelivr.net/qtip2/2.2.1/jquery.qtip.min.css) ).  
3. In Chrome, go to Tools > Extensions and press "Load unpacked extension." Point it to the directory containing all the files.  
4. You should now see the logo in the upper right (#), and inserted into every password field on any page you load.  

## How to use

1. Download & Install.  
2. Log in to your favorite website and find the "change password" webform.  
3. (Known Bug) When the password fields appear, they may not have the (#) logo in them. Click on the logo in the upper right and they should appear.  
4. In the "old password" field, type in your current password.  
5. In the "new password" and "confirm password" fields, click on the (#) logo. Alternatively, click on or tab into the text field and press Alt+H. A window should appear.  
6. Enter a passphrase, preferably around 20 characters long. (For a guide to choosing strong passphrases, click [here](http://xkcd.com/936/) )  
7. As you type, random-looking letters, numbers, and symbols will appear in the window. This is the hash code corresponding to what you typed.  
8. After typing your passphrase, click on the hashcode below (or press enter). The hashed password will be inserted into the password field and the window will disappear.  
9. Once you have changed your password, log out, and log in with the new password by repeating steps 5-8.  

## How it works

The app is powered by the SHA256 algorithm, implemented in Javascript. Credit: Angel Marin, Paul Johnston. On each keyup event, the app hashes "site | passphrase", so for example, "facebook | hello world" hashes to "GzLIi/lxeLaI8hMx" while "google | hello world" hashes to "etsemL+7ZSwFtX82". Therefore, the same passphrase can be used for multiple sites.  

Because there is no account, user id, password, or permission associated with the app, users can opt to use the same passphrase for every site, different passphrases, as complex or as simple a passphrase as they choose.  

## FAQ

1. Is it portable between devices? This is a *Chrome* plugin intended for use on any device capable of running Chrome. That said, our beta testers are hard at work answering this question.  

2. What if I'm on somebody else's computer, at work, or on a computer that's not running Chrome? Go to the [Product Page](http://www.pass-hash.com) and enter "site | passphrase" exactly as it always appears in the plugin. When you click on the hash a popup will appear prompting you to copy it to your clibboard.  

3. Can this run on any other browser other than Chrome? We're working on a Firefox plugin next.  

4. What about Internet Explorer? I'm a professional developer. Don't talk to me about IE.  

5. What is a hash function? A hash function, otherwise known as a one-way, or trapdoor function, is an algorithm that takes a string of text as input and very quickly and easily converts it to a string of random-looking output text. Given the output text, however, the input text is nearly impossible to decode. It's random-looking, not *random*, because the same input text *always* returns the same output text.  

6. Why SHA256? The Secure Hashing Algorithm SHA256 is one of many, many cryptographic hash functions that hasn't been broken yet. (MD5 and SHA1 have.)  

7. Could somebody break SHA256 then? Yes. They will eventually. At which point we'll switch to another standard.  

8. How will I know when the hackers have broken it? Let's not split hairs here. It took decades to render MD5 and SHA1 obsolete. There is no sudden safe vs. broken transition point. It's just an industry trend: the longer it's been in use, the less secure it is.  

9. Why is this amazing tool free? [shrugs]

changelog.md
Change Log
==========

### Version 1.1
#### 2014-09-30

- Favicon services are unreliable. I'm scraping the DOM for rel="icon" and using favicon service as a fallback (ex: yahoo) {'contentscript.js','popup.js'}  

- Tapped the "on visible" qtip event to refocus the phrase field if the user exits, then enters again. Couldn't directly call iframe.element.focus() because of a security restriction. Did not want to circumvent the iframe, so I found the "post message" [solution](http://stackoverflow.com/questions/14014167/how-can-i-pass-mouse-events-from-parent-to-child-iframe). {'contentscript.js','popup.js'}  

- I don't want to *force* the user to use a 10 character passphrase, because it confuses the "refresh on key up" animation. What I did do was make the text red for < 10 chars, then black for >= 10 chars. {'popup.js'}  

- Added hints. Password field tooltip becomes "PassHash (#) Press [Alt]+H" on focus, blank on blur. Clickable hash string tooltip is now "Press [Enter] to Insert". {'contentscript.js','popup.html'}  

- Sweep() function does not catch fields that are either invisible at page load or are loaded dynamically (ajax, etc). Ex: Facebook. When I go to "settings" and "change password" *new* fields appear, necessitating sweep() to run again. Still can't figure out how to trigger an actual event that would work for *all* sites. Instead, I'm rerunning sweep() on every *mouse click*. This might prove unnecessary and expensive, but certainly it is not as bad as runing it once every few seconds or something. Now, when you click *any* button, but incidentally the "change password" button, sweep() runs again. It *will still fail* b/c ajax is retrieving the pass field html, and no telling exactly how long to set timeout before it loads. But user will always *click* inside the password field to proceed, and then the (#) logo will appear. This will cause confusion, but it saves a step having to click the upper right icon. {'contentscript.js'}  

### Version 1.2
#### 2014-10-03

- Stripped down [typetype](https://github.com/iamdanfox/typetype) to insert text letter by letter. Neat effect, but this fixes a real bug with apple.com. Alt+H was triggering the password validator, and a straight elem.val() was not, leaving a "password is missing" error after inserting the password. This plugin simulates human typing. {'contentscript.js'}  

- Restore focus to active element after sweep() function runs. When password field appears dynamically, uniconified, and user clicks on field to focus it, sweep() runs (1.1). After sweep() runs though, focus was getting lost. {'contentscript.js'}  

- Two example sites, apple and dropbox, have password fields with a placeholder implemented as a styled "label for" rather than the "placeholder" attribute. Elem.val() was not clearing this, I think because these sites were using scripts to manage the placeholder. However, other sites use a "label for" password *correctly* as an actual label, that I don't want to hide. So, in addition to checking for its existence, I also have to check that it's .offset() left & top matches the password field. Ugly code. {'contentscript.js'}  

- Masked passphrase entry with a show/hide toggle, Alt+S. {'popup.js', 'popup.html', 'manifest.json'}  

- [Esc] key closes qtip. Beware of control flow problem! Qtip.hide() was calling popup.init(), leaving password field without focus. The *last* thing to be done in any series of callbacks is to hide the qtip, and let its callback function restore focus to the password field. {'contentscript.js'}  

- Changed message relay to relay whole message, no logic inside. It just receives a "request" object from popup.js and passes it down to contentscript.js. {'background.js'}  

### Version 1.3
#### 2015-01-12

- Added tab key to list of triggers for sweep(). An uniconified field could appear, and user will naturally either click on it or tab into it to enter a password. Either should trigger sweep() to iconify. Still looking for a more satisfactory solution here, without running a looping background monitor. {'contentscript.js'}  

- Changed text length from 24 to 23 b/c of Windows bug pushing icon to next line. {'popup.js'}  

###Version 1.6
####2015-06-22

- Refactored code such that chrome extension files could be used by a simple index.html file so I can host a pass-hash.com website without any repeated code. 

sha256.js

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
/**
*
*  Secure Hash Algorithm (SHA256)
*  http://www.webtoolkit.info/
*
*  Original code by Angel Marin, Paul Johnston.
*
**/
 
function SHA256(s){
 
    var chrsz   = 8;
    var hexcase = 0;
 
    function safe_add (x, y) {
        var lsw = (x & 0xFFFF) + (y & 0xFFFF);
        var msw = (x >> 16) + (y >> 16) + (lsw >> 16);
        return (msw << 16) | (lsw & 0xFFFF);
    }
 
    function S (X, n) { return ( X >>> n ) | (X << (32 - n)); }
    function R (X, n) { return ( X >>> n ); }
    function Ch(x, y, z) { return ((x & y) ^ ((~x) & z)); }
    function Maj(x, y, z) { return ((x & y) ^ (x & z) ^ (y & z)); }
    function Sigma0256(x) { return (S(x, 2) ^ S(x, 13) ^ S(x, 22)); }
    function Sigma1256(x) { return (S(x, 6) ^ S(x, 11) ^ S(x, 25)); }
    function Gamma0256(x) { return (S(x, 7) ^ S(x, 18) ^ R(x, 3)); }
    function Gamma1256(x) { return (S(x, 17) ^ S(x, 19) ^ R(x, 10)); }
 
    function core_sha256 (m, l) {
        var K = new Array(0x428A2F98, 0x71374491, 0xB5C0FBCF, 0xE9B5DBA5, 0x3956C25B, 0x59F111F1, 0x923F82A4, 0xAB1C5ED5, 0xD807AA98, 0x12835B01, 0x243185BE, 0x550C7DC3, 0x72BE5D74, 0x80DEB1FE, 0x9BDC06A7, 0xC19BF174, 0xE49B69C1, 0xEFBE4786, 0xFC19DC6, 0x240CA1CC, 0x2DE92C6F, 0x4A7484AA, 0x5CB0A9DC, 0x76F988DA, 0x983E5152, 0xA831C66D, 0xB00327C8, 0xBF597FC7, 0xC6E00BF3, 0xD5A79147, 0x6CA6351, 0x14292967, 0x27B70A85, 0x2E1B2138, 0x4D2C6DFC, 0x53380D13, 0x650A7354, 0x766A0ABB, 0x81C2C92E, 0x92722C85, 0xA2BFE8A1, 0xA81A664B, 0xC24B8B70, 0xC76C51A3, 0xD192E819, 0xD6990624, 0xF40E3585, 0x106AA070, 0x19A4C116, 0x1E376C08, 0x2748774C, 0x34B0BCB5, 0x391C0CB3, 0x4ED8AA4A, 0x5B9CCA4F, 0x682E6FF3, 0x748F82EE, 0x78A5636F, 0x84C87814, 0x8CC70208, 0x90BEFFFA, 0xA4506CEB, 0xBEF9A3F7, 0xC67178F2);
        var HASH = new Array(0x6A09E667, 0xBB67AE85, 0x3C6EF372, 0xA54FF53A, 0x510E527F, 0x9B05688C, 0x1F83D9AB, 0x5BE0CD19);
        var W = new Array(64);
        var a, b, c, d, e, f, g, h, i, j;
        var T1, T2;
 
        m[l >> 5] |= 0x80 << (24 - l % 32);
        m[((l + 64 >> 9) << 4) + 15] = l;
 
        for ( var i = 0; i<m.length; i+=16 ) {
            a = HASH[0];
            b = HASH[1];
            c = HASH[2];
            d = HASH[3];
            e = HASH[4];
            f = HASH[5];
            g = HASH[6];
            h = HASH[7];
 
            for ( var j = 0; j<64; j++) {
                if (j < 16) W[j] = m[j + i];
                else W[j] = safe_add(safe_add(safe_add(Gamma1256(W[j - 2]), W[j - 7]), Gamma0256(W[j - 15])), W[j - 16]);
 
                T1 = safe_add(safe_add(safe_add(safe_add(h, Sigma1256(e)), Ch(e, f, g)), K[j]), W[j]);
                T2 = safe_add(Sigma0256(a), Maj(a, b, c));
 
                h = g;
                g = f;
                f = e;
                e = safe_add(d, T1);
                d = c;
                c = b;
                b = a;
                a = safe_add(T1, T2);
            }
 
            HASH[0] = safe_add(a, HASH[0]);
            HASH[1] = safe_add(b, HASH[1]);
            HASH[2] = safe_add(c, HASH[2]);
            HASH[3] = safe_add(d, HASH[3]);
            HASH[4] = safe_add(e, HASH[4]);
            HASH[5] = safe_add(f, HASH[5]);
            HASH[6] = safe_add(g, HASH[6]);
            HASH[7] = safe_add(h, HASH[7]);
        }
        return HASH;
    }
 
    function str2binb (str) {
        var bin = Array();
        var mask = (1 << chrsz) - 1;
        for(var i = 0; i < str.length * chrsz; i += chrsz) {
            bin[i>>5] |= (str.charCodeAt(i / chrsz) & mask) << (24 - i%32);
        }
        return bin;
    }
 
    function Utf8Encode(string) {
        string = string.replace(/\r\n/g,"\n");
        var utftext = "";
 
        for (var n = 0; n < string.length; n++) {
 
            var c = string.charCodeAt(n);
 
            if (c < 128) {
                utftext += String.fromCharCode(c);
            }
            else if((c > 127) && (c < 2048)) {
                utftext += String.fromCharCode((c >> 6) | 192);
                utftext += String.fromCharCode((c & 63) | 128);
            }
            else {
                utftext += String.fromCharCode((c >> 12) | 224);
                utftext += String.fromCharCode(((c >> 6) & 63) | 128);
                utftext += String.fromCharCode((c & 63) | 128);
            }
 
        }
 
        return utftext;
    }
 
    function binb2hex (binarray) {
        var hex_tab = hexcase ? "0123456789ABCDEF" : "0123456789abcdef";
        var str = "";
        for(var i = 0; i < binarray.length * 4; i++) {
            str += hex_tab.charAt((binarray[i>>2] >> ((3 - i%4)*8+4)) & 0xF) +
            hex_tab.charAt((binarray[i>>2] >> ((3 - i%4)*8  )) & 0xF);
        }
        return str;
    }
 
    s = Utf8Encode(s);
    return binb2hex(core_sha256(str2binb(s), s.length * chrsz));
 
}

chrome/popup.js

/**
 * PassHash > popup.js
 * Written by Victor Davis Sep 2014
 * Contact VictorAlynDavis@gmail.com
 * 
 * This is the file that pops up when the user activates the extension to interact with it.
 */


var pass = {
  /**
   * Converts hexadecimal to base64
	 * Credit: http://stackoverflow.com/questions/23190056/hex-to-base64-converter-for-javascript
   */
  HexToBase64: function(string) {
		return btoa(string.match(/\w{2}/g).map(function(a){return String.fromCharCode(parseInt(a,16));}).join(""));
	},
	
  /**
   * The algorithm:
	 * (1) Append the site & phrase with a vertical bar, like "facebook | hello world"
	 * (2) Hash using SHA256 (separate downloaded file)
	 * (3) Convert hexed output to base64
	 * (4) Use only the first 16 characters (some websites set upper limits on length)
	 * Results in a 64x16= 1024-bit password
   */
  hash: function() {
		// get phrase & site
    var phrase = document.getElementById("hash_phrase").value;
    var site = document.getElementById("hash_site").value;
		var plaintext = site + ' | ' + phrase;
		
		// 1.2 don't do any work if the field is empty
		if ((phrase.length > 0) | (site.length > 0)) {
			// show show/hide icon
			document.getElementById('showhide').style.display = 'inline';
		
			// 1.1 extra special security: encourage passphrase to be
			// longer than 10 characters
			phrase_len = phrase.length;
			bg = "";
			if (phrase_len < 10) {
				bg = "red";
			} else {
				bg = "black";
			}
			document.getElementById('hash_result').style.color = bg;
			
			// hash away
			var ciphertext = SHA256(plaintext);
			ciphertext = this.HexToBase64(ciphertext);
			ciphertext = ciphertext.substring(0,16);
			
			// display result
			document.getElementById('hash_result').innerHTML = ciphertext;
		} else {
			// hide show/hide icon, clear result
			document.getElementById('showhide').style.display = 'none';
			document.getElementById('hash_result').innerHTML = '';
		}
  },
  
	/**
	 * When user clicks on the result, send ciphertext down to the content script,
	 * and initialize all inputs
	 */
  insert: function(word) {
		var msg = { action: 'insert_text', pass: word };
		if (word.length > 0) {
			if ((typeof(chrome) != "undefined") && (typeof(chrome.extension) != "undefined"))
				chrome.extension.sendMessage(msg, function(response){});
			else
				window.prompt("Copy to clipboard: Ctrl+C, Enter", word);
		}
	},

	/**
	* short utility functions
	*/
	toggle: function() {
		var type = document.getElementById('hash_phrase').type;
		
		var show = (type == 'password');
		var fieldtype = show ? 'text': 'password';
		var imgsrc    = show ? 'hide20.png': 'show20.png';
		var imgtitle  = show ? 'Hide': 'Show';
		
		document.getElementById('hash_phrase').type = fieldtype;
		document.getElementById('showhide').src = imgsrc;
		document.getElementById('showhide').title = imgtitle+' Passphrase Alt+S';
	},
	init: function(which) {
		this.toggle('hide');
		document.getElementById('hash_result').innerHTML = '';
		document.getElementById('hash_phrase').value = '';
		document.getElementById('showhide').style.display = 'none';
		
		if (which == 'visible') {
			document.addEventListener('keydown', pass.listen);
			if (document.getElementById('hash_site').disabled)
				document.getElementById('hash_phrase').focus();
			else
				document.getElementById('hash_site').focus();
		} else {
			// 1.2 remove listeners!
			document.removeEventListener('keydown', pass.listen);
		}
	},
	
	/*
	* key listener
	* 1.2 make sure to disable on qtip hide
	*/
	listen: function(e) {
		var key = e.keyCode;
		if (key == 13) { /* enter */
			var word = document.getElementById("hash_result").innerHTML;
			pass.insert(word);
		} else if (e.altKey & (key == 83)) { /* alt+s */
			pass.toggle();
		} else if (key == 27) { /* esc */
			// 1.2 tell content script to close qtip
			if ((typeof(chrome) != "undefined") && (typeof(chrome.extension) != "undefined")) {
				chrome.extension.sendMessage(
					{ action: 'close' },
					function(response){}
				);
			}
		}
	}
};

fill_site = function (result) {
	var hash_site = document.getElementById('hash_site');
	var hash_phrase = document.getElementById('hash_phrase');
	var hash_pipe = document.getElementById('hash_pipe');

	//console.log(result);
	site = result.site;
	domain = result.domain;
	icon = result.icon;

	// populate inputs
	if (result.site) {
		hash_site.value = site;
		hash_site.size = site.length;
		hash_phrase.size = 19 - site.length;
	} else {
		hash_site.size = 8;
		hash_phrase.size = 11;
	}

	// NONE OF THESE WORK RELIABLY
	// https://www.google.com/s2/favicons?domain=www.[] misses github
	// https://plus.google.com/_/favicon?domain=www.[] misses twitter
	// https://plus.google.com/_/favicon?domain=[] misses dropbox
	// http://g.etfv.co/http://www.[] appears to be defunct
	if (icon == "") {
		icon = "http://www.google.com/s2/favicons?domain=http://www."+domain;
	}
	hash_site.style.backgroundImage = "url('"+icon+"')";
	hash_site.style.backgroundRepeat = "no-repeat";
	hash_site.style.backgroundSize = "contain";
	hash_site.style.paddingLeft = "20px";
}

/**
 * On load, add listeners
 * 
 */
document.addEventListener('DOMContentLoaded', function() {
	
	// refresh output at every keystroke, the algorithm is plenty fast, no need for a "submit" button
	var hash_click = document.getElementById('hash_phrase');
	hash_click.addEventListener('keyup', function() {
		pass.hash();
	});
	var hash_site = document.getElementById('hash_site');
	hash_site.addEventListener('keyup', function() {
		pass.hash();
	});

	// tab out of site triggers icon lookup
	hash_site.addEventListener('blur', function() {
		var info = { 'site': this.value, 'domain':this.value + '.com', 'icon':'' };
		fill_site(info);
	});

	// make the result text clickable, click to insert, again, saving a button
	var hash_insert = document.getElementById('hash_insert');
	hash_insert.addEventListener('click', function() {
		var word = document.getElementById('hash_result').innerHTML;
		pass.insert(word);
	});
	
	// retrieve website / domain ("facebook"/"facebook.com") from local storage
	if ((typeof(chrome) != "undefined") && (typeof(chrome.storage) != "undefined")) {
		chrome.storage.sync.get(['site','domain','icon'], fill_site);
	} else {
		var info = { 'site': '', 'domain':'', 'icon':'' };
		fill_site(info);
		hash_site.removeAttribute("disabled");
	}

	/**
	* Show/hide password toggle
	*/
	document.getElementById('showhide').addEventListener('click', function() {
		pass.toggle();
	});
	
	// autofocus passphrase
	pass.init();
});

/*
 * 1.1 If user enters, then exits, then enters qtip, refocus passphrase
 * (ondomloaded is catching this, but onshow can't reach inside the iframe
 *  to focus because of security restrictions -- use post message)
 */
window.addEventListener('message',function(e) {
	if (e.data.visible) { /* on qtip show */
		pass.init('visible');
	}
	if (e.data.hidden) { /* on qtip hide */
		pass.init('hidden');
	}
});


chrome/background.js

/**
 * PassHash > background.js
 * Written by Victor Davis Sep 2014
 * Contact VictorAlynDavis@gmail.com
 * 
 * This file is necessary to relay messages from content script to popup and vice versa,
 * it is always running in the background as long as the extension is enabled
 */

/**
 * Occurs when the user clicks the icon in the upper right.
 * Send message to content script to sweep the page for un-mutated password fields
 * Certain websites like facebook, homedepot, put a sign-in in a popup screen,
 * so the sweep() function did not catch them on DOM load.
 */
chrome.browserAction.onClicked.addListener(
	function() {
		// trigger content script to sweep for password inputs
		chrome.tabs.query({active: true, currentWindow: true}, function(tabs){
			var obj = {action: "sweep"};
			chrome.tabs.sendMessage(tabs[0].id, obj, function(response) {});
		});
	}
);

/**
 * Occurs when user clicks "insert" button.
 * Popup cannot access DOM or talk to content script directly, so this
 * listener runs in the background to relay the "insert_text" command from one to the other
 */
chrome.extension.onMessage.addListener(
	function(request, sender, sendResponse) {
		var word = request.pass;
		sendResponse({confirmation: word});
		
		// relay message on to content script
		chrome.tabs.query({active: true, currentWindow: true}, function(tabs){
			var obj = {action: "insert_text", pass: word};
			chrome.tabs.sendMessage(tabs[0].id, obj, function(response) {});
		});
	}
);


chrome/contentscript.js

/**
 * PassHash > contentscript.js
 * Written by Victor Davis Sep 2014
 * Contact VictorAlynDavis@gmail.com
 * 
 * This file contains all the code necessary for modifying the DOM.
 * Some sensitive JS and CSS tricks must be employed to modify 
 * a major website's login screen without breaking it.
 */

/**
 * Global variable holding the DOM element into which the hashed password
 * will be inserted. Bad form. Change this later.
 */
activePass = null;
html = false;

/*
 * extracts simple domain name, ignoring .com, .net, and subdomains www. etc
 * and sends it to background for inter-extension storage
 */
function saveDomain() {
	
	// ex facebook google homedepot, and write to background local storage
	var site = document.domain.toLowerCase().match("([a-z]*)\.[a-z]*$");
	if (!site) site = [];
	
	// get icon too (those favicon services just don't work
	icon = "";
	var link = document.getElementsByTagName("link");
	for (var i = 0; i < link.length; i++) {
		if ( (link[i].hasAttribute("rel"))
			&& (link[i].hasAttribute("href"))
			&& (link[i].rel.indexOf("icon") !== -1) ) {
			icon = link[i].href;
		}
	}
	
	//console.log('sending domain "'+site[0]+'" to background...');
	var info = { 'site': site[1], 'domain':site[0], 'icon':icon };
	chrome.storage.sync.set(
	  info,
	  function() {
		// Notify that we saved.
	});
}

/**
 * Runs on DOM load. Also runs when user clicks the icon in the upper right.
 * Scans the page for password input fields and iconifies each by inserting 
 * PassHash icon (#) in the far right. Clicking on the icon opens the popup,
 * implemented as a qtip: http://qtip2.com/
 * 
 * NOTE: QTip has a jQuery dependency, so it is this feature that requires
 * us to load jQuery library
 */
function sweep() {
	
	// 1.2 sweep() bumps focus off active element, restore manually
	var activeElem = document.activeElement;
	
	// scan all inputs for type="password" attribute
	var inputs = document.getElementsByTagName("input");
	var i, passlist = [];
	for (i = 0; i < inputs.length; i++) {
		if (inputs[i].type == "password") {
			passlist.push(inputs[i]);
		}
	}

	// for each result, iconify the field
	for (i = 0; i < passlist.length; i++) {
		
		// if this field has not *already* been passhashed
		var zero = passlist[i].parentElement.getElementsByClassName("passhash").length;
		
		// paypal loads password fields in the background, then displays them.
		// passhash should not mess with them unless they are visible
		var paypal = (passlist[i].offsetHeight > 0);
		
		// two prereqs: that the field is not *already* iconified and the field is visible
		if ((zero == 0) & paypal) {
		
			// create wrapper in which to insert input + image
			reladiv = document.createElement("div");
			reladiv.className = "reladiv";
			reladiv.style.position = "relative";
			reladiv.style.height = passlist[i].offsetHeight;
			reladiv.style.margin = '0px';
			reladiv.style.padding = '0px';
			
			// create clickable image to far right of field
			img = document.createElement("img");
			img.src = chrome.extension.getURL("icon128.png");
			img.className = "passhash";
			img.height = passlist[i].offsetHeight;
			img.style.height = passlist[i].offsetHeight+'px';
			img.style.position = "absolute";
			img.style.top = "0px";
			img.style.zIndex = 3;
			
			// appendChild will reorder your node to the end of its list of sisters, must use replaceChild
			myparent = passlist[i].parentElement;
			myparent.replaceChild(reladiv, passlist[i]);
			
			// insert password field + image into new parent div
			reladiv.appendChild(passlist[i]);
			reladiv.appendChild(img);
			
			// accurately set the positioning
			img.style.left = (passlist[i].offsetWidth + passlist[i].offsetLeft - passlist[i].offsetHeight)+"px";
			img.style.top = (passlist[i].offsetTop)+"px";
			
			// use Alt+H to open qtip
			passlist[i].addEventListener('keyup', function(e) {
				if (e.altKey & (e.keyCode == 72)) { /* alt+h */
					activePass = this;
					saveDomain();
					this.parentElement.getElementsByTagName("img")[0].click();
				}
			});
			
			// 1.1 give user hint about Alt+H
			passlist[i].addEventListener('focus', function() {
				this.parentElement.title = "PassHash (#) Press [Alt]+H";
			} );
			passlist[i].addEventListener('blur', function() {
				this.parentElement.title = "";
			} );
			
			// not a link! create click function for icon
			// When icon is clicked, write the website name to local storage,
			// NOTE: user may have multiple tabs open and they all share the extension,
			// so by clicking on an icon, the user is changing the domain
			img.onclick = function() {
				
				// register active password field as global variable (BAD FORM)
				activePass = this.parentElement.getElementsByTagName("input")[0];
				//console.log(activePass);
				
				// save domain name to background
				saveDomain();
			}
			
			// create content for the popup
			iframe = document.createElement("iframe");
			iframe.id = "ifpopup";
			iframe.src = chrome.extension.getURL("popup.html");
			iframe.height = "42px";
			iframe.width = "100%";
			iframe.style.borderWidth = "0";
			
			// qtip2.com allows for tooltips with html content, requires jQuery
			$('.passhash').qtip( 
			{
				content: $(iframe),
				position: {
					at: "bottom right",
					my: "top right"
				},
				show: {
					event: 'click' /* do not show tooltip on mouse over, wait for click */
				},
				hide: {
					event: 'nonexisting' /* wait for programmatic close */
				},
				events: {
					/* 1.1 post message -- can't reach into iframe.element.focus() per security */
					visible : function() {
						$(iframe)[0].contentWindow.postMessage(
							{ "visible": true }, "*"
						);
					},
					hidden : function() {
						$(iframe)[0].contentWindow.postMessage(
							{ "hidden": true }, "*"
						);
						$(window.activePass).focus();
					}
				},
				style: {
					width: "400px",
					"max-width": "400px",
					"min-width": "400px",
				}
			});
		}
	}
	
	// 1.2 restore focus to active element
	activeElem.focus();
}

/**
 * 1.2
 * Modified code from "typetype" plugin, inserts text into input one character at
 * a time using a recursive queued callback, which magically works for apple.com
 * 
 * https://github.com/iamdanfox/typetype
 */
$.fn.extend({
	typetype: function(txt, options) {
		var settings;
		
		settings = $.extend({
			callback: function() {}, /* runs after text insert */
			keypress: function() {}  /* runs after each keystroke */
		}, options);
		
		return this.each(function() {
			var elem;
			
			// this is necessary, can't just find/replace elem with "this"
			elem = this;
			
			return $(elem).queue(function() {
				var typeTo;
				
				typeTo = function(i) {
					
					if (i <= txt.length) {
						elem.value += txt[i - 1];
						settings.keypress.call(elem);
						
						setTimeout((function() {
							return typeTo(i + 1);
						}), 0);
						
					} else {
						settings.callback.call(elem);
						$(elem).dequeue();
					}
					
				};
				
				return typeTo(1);
			});
		});
	}
});

/**
 * Run sweep() function on DOM load
 * 1.1 Also on every mouse click, to catch the hidden fields
 */
$(document).ready(function() {
	sweep();
	document.body.addEventListener('click',
		function() {
			sweep();
		}
	);

	document.body.onclick = function(e) {
		if (!html) $('.passhash').qtip("hide");
	};
	chrome.extension.onMessage.addListener(function(msg, sender, sendResponse) {
	
		/**
		 * weird bug on ebay -- content script runs once per iframe on the page,
		 * so insert() sends once, but onMessage receives twice, only once with 
		 * access to window.activePass
		 */
		if (window.activePass == null) {
			console.log({'content script running in iframe, src = ': document.URL});
			return;
		}
	
		// sent from popup pass.insert()
	  if (msg.action == 'insert_text') {
			/** 1.2 get rid of "label for" placeholder (apple, dropbox)
			 * CAREFUL! Some websites use "label for" the *right* way, as a *label*
			 * that says "Password:" not as a *placeholder* -- to distinguish,
			 * placeholder's offset().left is the same as the field itself
			 * ex: dropbox has <label>password
			 * ex: but apple has <label><span>password
			 * So check children's offset too! argh ><
			 */
			var label = document.getElementsByTagName("label");
			for (var i = 0; i < label.length; i++) {
				if ( (label[i].hasAttribute("for"))
					&& (label[i].attributes.for.value == activePass.id)
				  && ( overlapping($(window.activePass), $(label[i])) ) ) {
					label[i].remove();
				}
			}
		
			// disappear qtip after insert
			//if (!html) $('.passhash').qtip("hide");
			$(activePass).focus();
			$(activePass).val('');
		
			/**
			 * 1,2 asynchronous call to modified typetype, yields neat effect
			 * of human typing, but more than that, this hodgepodge of calls
			 * actually fixes apple.com bug where validator doesn't trigger
			 */
			$(activePass).typetype(msg.pass, {
				callback: function() {
					// apple.com after inserting text, the password validator
					// does not get triggered until you do all this mess
					this.blur();
					this.focus();
				
					// make sure to hide qtip *last* -- its callback restores focus
					// to active password field
					if (!html) $('.passhash').qtip("hide");
				}
			});
		
	  } else if (msg.action == 'sweep') {
			sweep();
		} else if (msg.action == 'close') {
			if (!html) $('.passhash').qtip("hide");
		}
	});
	if (html) document.getElementsByTagName("img")[0].click();
});

/*
 * 1.2 Can't believe I have to do this, argh.
 * Some password fields have a "placeholder" attribute that disappears automatically
 * when user enters text (correct). Others have a "label for" whose visibility the site
 * controls via script (incorrect). Can't disappear "label for" automatically because
 * yet other sites use it (correctly) as an actual label. Disappear it *only* if it
 * physically overlaps the password field. TODO: find a way to fool the scripts
 * ex: disappear on apple & dropbox, leave on facebook & github & coursera
 */
function overlapping(a, b) {
	var ret = false;
	var aoff = a.offset();
	var boff = b.offset();
	
	// lefts match, tops match within +/- 5px
	if ( (aoff.left == boff.left)
		&& (aoff.top - boff.top < 5) 
		&& (boff.top - aoff.top < 5) ) {
		ret = true;
	}
	
	// run same test on children -- recursive
	if (b.children().length > 0) {
		ret = ret || overlapping(a, b.children(":first"));
	}
	
	return ret;
}

/**
 * Listens for "insert" command from popup, and inserts argument into active password field
 */
